op-batcher/batcher/espresso/streamer.go

```diff
@@ -0,0 +1,225 @@
+	s.Log.Debug("Updated finalized hotshot state", "hotshotState", hotshotState)
+
+	targetHeight := min(hotshotState.BlockHeight, s.hotShotPos+100)
+
+	for ; s.hotShotPos < targetHeight; s.hotShotPos += 1 {
+		s.Log.Debug("fetching HotShot block", "blockNr", s.hotShotPos)
+
+		txns, err := s.EspressoClient.FetchTransactionsInBlock(ctx, s.hotShotPos, s.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to fetch transactions in block: %w", err)
+		}
+
+		if len(txns.Transactions) == 0 {
+			s.Log.Debug("no transactions in hotshot block", "blockNr", s.hotShotPos)
+			continue
+		}
+
+		rawHeader, err := s.EspressoClient.FetchRawHeaderByHeight(ctx, s.hotShotPos)
+		if err != nil {
+			return fmt.Errorf("failed to fetch raw HotShot header: %w", err)
+		}
+
+		var header espressoTypes.HeaderImpl
+		err = json.Unmarshal(rawHeader, &header)
+		if err != nil {
+			return fmt.Errorf("could not unmarshal header from bytes")
+		}
+
+		snapshot, err := s.EspressoLightClient.FetchMerkleRoot(s.hotShotPos, nil)
+		if err != nil {
+			return fmt.Errorf("failed to fetch Merkle root: %w", err)
+		}
+
+		if snapshot.Height <= s.hotShotPos {
+			return fmt.Errorf("snapshot height is less than or equal to the requested height")
+		}
+
+		nextHeader, err := s.EspressoClient.FetchHeaderByHeight(ctx, snapshot.Height)
+		if err != nil {
+			return fmt.Errorf("error fetching the snapshot header (height: %d): %w", snapshot.Height, err)
+		}
+
+		proof, err := s.EspressoClient.FetchBlockMerkleProof(ctx, snapshot.Height, s.hotShotPos)
+		if err != nil {
+			return fmt.Errorf("error fetching merkle proof")
+		}
+
+		blockMerkleTreeRoot := nextHeader.Header.GetBlockMerkleTreeRoot()
+
+		log.Info("Verifying merkle proof", "height", s.hotShotPos)
+		ok := espressoVerification.VerifyMerkleProof(proof.Proof, rawHeader, *blockMerkleTreeRoot, snapshot.Root)
+		if !ok {
+			return fmt.Errorf("error validating merkle proof (height: %d, snapshot height: %d)", s.hotShotPos, snapshot.Height)
+		}
+
+		namespaceOk := espressoVerification.VerifyNamespace(
+			s.Namespace,
+			txns.Proof,
+			*header.Header.GetPayloadCommitment(),
+			*header.Header.GetNsTable(),
+			txns.Transactions,
+			txns.VidCommon,
+		)
+
+		if !namespaceOk {
+			s.Log.Error("namespace verification failed for HS block", "blockNr", s.hotShotPos)
+			return fmt.Errorf("namespace verification failed")
+		}
+
+		for _, transaction := range txns.Transactions {
+			batch, err := UnmarshalEspressoTransaction(transaction, s.BatcherAddress)
+			if err != nil {
+				s.Log.Info("Failed to unmarshal espresso transaction", "error", err)
+				continue
+			}
+
+			if batch.BlockNum < s.BatchPos {
+				// Batch already buffered/finalized
+				s.Log.Debug("batch is older than current batchPos, skipping", "batchNr", batch.BlockNum, "batchPos", s.BatchPos)
+				continue
+			}
+
+			espressoFinalizedL1 := getFinalizedL1(&header)
+			if espressoFinalizedL1 == nil {
+				s.Log.Error("Unknown Espresso header version")
+				continue
```

Comment:
IMO this should be an error, not a `continue`. The Espresso header version cannot be faked as it's part of the onchain data authentiated by Espresso consensus. Thus, if this happens, it means Espresso has done an upgrade that this version of the batch poster (or caffeinated node, if sharing a streamer) is not prepared to handle. In that case we should probably just error out and force the operator to deploy a new version, rather than continuing and possibly deriving an incorrect state

Reply:
Fair, I'll update that.